packages/core/clockworker.js

```diff
@@ -0,0 +1,128 @@
+function getTime(precision) {
+  const seconds = performance.now() / 1000;
+  return Math.round(seconds * precision) / precision;
+}
+
+let numPorts = 0;
+let num_cycles_at_cps_change = 0;
+let num_ticks_since_cps_change = 0;
+let cps = 0.5;
+const duration = 0.1;
+const channel = new BroadcastChannel('strudeltick');
+
+const sendMessage = (type, payload) => {
+  channel.postMessage({ type, payload });
+};
+
+const sendTick = ({ phase, duration, time }) => {
+  sendMessage('tick', {
+    phase,
+    duration,
+    time,
+    cps,
+    num_cycles_at_cps_change,
+    num_ticks_since_cps_change,
+  });
+  num_ticks_since_cps_change++;
+};
+
+const clock = createClock(sendTick, duration);
+let started = false;
+
+const startClock = () => {
+  if (started) {
+    return;
+  }
+  clock.start();
+  started = true;
+};
+const stopClock = async () => {
+  //dont stop the clock if mutliple instances are using it...
+  if (!started || numPorts !== 1) {
+    return;
+  }
+  clock.stop();
+  setCycle(0);
+  started = false;
+};
+
+const setCycle = (cycle) => {
+  num_ticks_since_cps_change = 0;
+  num_cycles_at_cps_change = cycle;
+};
+
+const processMessage = (message) => {
+  const { type, payload } = message;
+
+  switch (type) {
+    case 'cpschange': {
+      if (payload.cps !== cps) {
+        num_cycles_at_cps_change = num_cycles_at_cps_change + num_ticks_since_cps_change * duration * cps;
+        cps = payload.cps;
+        num_ticks_since_cps_change = 0;
+      }
+      break;
+    }
+    case 'setcycle': {
+      setCycle(payload.cycle);
+      break;
+    }
+    case 'toggle': {
+      if (payload.started) {
+        startClock();
+      } else {
+        stopClock();
+      }
+      break;
+    }
+  }
+};
+
+this.onconnect = function (e) {
+  // the incoming port
+  const port = e.ports[0];
+  numPorts = numPorts + 1;
+  port.addEventListener('message', function (e) {
+    processMessage(e.data);
+  });
+  port.start(); // Required when using addEventListener. Otherwise called implicitly by onmessage setter.
+};
+
+function createClock(
```

Comment:
why has this been pulled out? i don't see any real difference between the old zyklus.mjs and this function, except some inputs being a closure now, also I don't understand why we need to round twice, once in getTime and once more in the while loop

Reply:
hmmm so I tried adding the import for zyklus, and it worked in chrome but not in firefox for me.

Reply:
okay so it looks like importScripts works.It is supported for everything but deno it looks like https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts